### ISSUE TITLE
Question about training

### ISSUE THREAD

**Issue by YuanLiuuuuuu** [neuroailab/VIE#19]
Hi, I have another two questions to ask you. First, just as this line https://github.com/neuroailab/VIE/blob/db9cd758947dd9a3e16701e24f9716df437792c9/tf_model/train_vie.py#L208 shows, do you first update your memory bank and then calculate your loss for current batch ? Second, do you use back propagation to update your memory bank during your training?

**Comment by chengxuz** [comment]
The memory bank is updated after its value is used to compute the loss and the gradients, so the loss and the gradients are computed using old values in the memory bank, instead of the newly updated value. Back propagation is not used to update the memory bank, it is only updated by the 0.5 old and new embedding mix done at: https://github.com/neuroailab/VIE/blob/db9cd758947dd9a3e16701e24f9716df437792c9/tf_model/model/instance_model.py#L210.

**Comment by YuanLiuuuuuu** [comment]
However, tf.control_dependencies(update_ops) clearly updates memory bank before we use it to compute loss and gradients. I run a experiment, just like the code bellow；
`label = tf.placeholder(tf.float32, (None, 4))
mb = tf.get_variable("memory_bank", initializer=tf.ones((10, 4), dtype=tf.float32), dtype=tf.float32)

index = tf.constant([0, 1])
y = tf.gather(mb, index, axis=0)

update_tensor = tf.zeros((2, 4), dtype=tf.float32)
update_index = tf.constant([0, 1], dtype=tf.int32)
update_mb = tf.scatter_update(mb, update_index, update_tensor)

res = y - label

sess = tf.Session()
sess.run(tf.global_variables_initializer())

data = np.ones((2, 4), dtype=np.float32)

with tf.control_dependencies([update_mb]):
         res = tf.identity(res)
data_res = sess.run(res, feed_dict={label: data})
mb_data = sess.run(mb)`

data_res is the value after I update memory bank. I think your code follows the same paradigm.

**Comment by YuanLiuuuuuu** [comment]
Hi, Cheng, hope you can help me.

**Comment by chengxuz** [comment]
Hi, I saw your email on your message before your editting, so I sent a zoom link to your email address, but I think you edited your message to remove that, is that email address on your earlier message valid?

About your question, I have run your codes and validated that the data_res is the value before the memory bank update (all zeros), you should notice that res is y-label, it's not just y.

**Comment by YuanLiuuuuuu** [comment]
How many steps did you train resnet3d model on kinetics?

**Comment by chengxuz** [comment]
You can check the logs in the ipython notebook, the validation is done every 5000 steps. So the total number of steps can be computed through multiplying the largest number in the x-axis by 5000.

**Comment by YuanLiuuuuuu** [comment]
From your ipython notebook, you trained for 600,000 steps，however, in your code, you set the lr decay steps to 1750011,1924998. You did not use it.

**Comment by chengxuz** [comment]
Where did you see the lr decay steps as `1750011,1924998`? At https://github.com/neuroailab/VIE/blob/master/tf_model/saved_settings/vd_3dresnet_fx.py#L42, it clearly says that the lr decay is at `974946,1304998`. Besides, in the ipython notebook, the first figure in the notebook shows that the maximal number on the x-axis is around 300, which means the maximal number of steps is around 1.500,000. 

Are you talking about transfer learning? If that's the case, then it's because the global_step (step number) was loaded from the pretrained model in the transfer learning, which is why the lr decay for transfer learning tends to be very big.

**Comment by YuanLiuuuuuu** [comment]
Thanks
